### PR TITLE
Update Temperature Measurement gen/ folder from ZAP

### PR DESCRIPTION
**Pull request by vivien-apple**
#### Problem

#3464 is trying to update all the examples in once. This PR upgrade only the `temperature-measurement-app`.

In order to produce it, I did a fork with some changes to the ZAP templates (#3780, #3809, #3810 and #3824) and with an extra change to `src/app/util/util.cpp` (#3825) as I forgot to add a `#ifdef` at some point.

In order to land, the only requirement from those 5 PRs is #3825.

Then I ran: `node src-script/zap-generate.js -z ./zcl-builtin/silabs/zcl.json -g ../../app/zap-templates/chip-templates.json -i temperature.zap -o ../../../examples/temperature-measurement-app/esp32/main/gen`

`temperature.zap` is a slightly modified version from the version in #3464 since the version in #3464  kills the commands of the Basic Cluster.

It will require a bit more additional work for the other examples app, notably because the current template for `gen_config.h` is missing some `#define`, but in the case of the `temperature-measurement-app` enough changes has landed inside CHIP that we can start to use the template from ZAP.

 #### Summary of Changes
  * Replace most of the `gen/` files from the `temperature-measurement-app`

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

**Review by jelderton — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]

File: examples/temperature-measurement-app/esp32/main/gen/callback-stub.cpp
Comment (vivien-apple) [reply examples/temperature-measurement-app/esp32/main/gen/callback-stub.cpp:74]: I opened #3846 for that. I don't mind waiting for #3846 to land before this one lands it we want to make the history a little bit cleaner.

File: examples/temperature-measurement-app/esp32/main/gen/af-structs.h
Comment (vivien-apple) [reply examples/temperature-measurement-app/esp32/main/gen/af-structs.h:1]: You should probably not take it on faith.

If you look at it twice you may noticed that there are some empty structs. For example:

With `ZAP` templates:
```
// Struct for Signature
typedef struct _Signature
{
} EmberAfSignature;
```

With `AppBuilder`:
```
// Void typedef for EphemeralData which is empty.
// this will result in all the references to the data being as uint8_t*
typedef uint8_t EphemeralData;
```

It does work because those are not used. I will open an issue for them.

File: examples/temperature-measurement-app/esp32/main/gen/call-command-handler.cpp
Comment (vivien-apple) [reply examples/temperature-measurement-app/esp32/main/gen/call-command-handler.cpp:105]: That's a lack of good way to express that with the templates helper. I will open an issue to track that.

**Review by BroderickCarlin — APPROVED** [review]

**Comment by BroderickCarlin** [comment]
@vivien-apple can you address the merge conflicts? Should be good to go after that!

**Comment by vivien-apple** [comment]
> @vivien-apple can you address the merge conflicts? Should be good to go after that!

Thanks. I'm waiting a little bit, secretly hoping that #3846 lands before ;) If it does not land soon not I will rebase this one.

**Review by rwalker-apple — APPROVED** [review]

File: examples/temperature-measurement-app/esp32/main/gen/af-structs.h
Comment (vivien-apple) [reply examples/temperature-measurement-app/esp32/main/gen/af-structs.h:1]: #3873

File: examples/temperature-measurement-app/esp32/main/gen/call-command-handler.cpp
Comment (vivien-apple) [reply examples/temperature-measurement-app/esp32/main/gen/call-command-handler.cpp:107]: #3874

**Review by mspang — APPROVED** [review]

**Pull request merged**
